traits: restructure container type traits

mzuzek opened the pull request:
This PR is a new version of container type traits refactoring started on #359
driven by requirements and conclusions specified in https://github.com/Pressio/pressio/issues/356#issuecomment-1234062162.

mzuzek commented on include/pressio/ops/eigen/ops_level2.hpp:
```diff
@@ -64,9 +64,12 @@ template <
-  && ::pressio::Traits<A_type>::package_identifier == PackageIdentifier::Eigen
-  && ::pressio::Traits<x_type>::package_identifier == PackageIdentifier::Eigen
-  && ::pressio::Traits<y_type>::package_identifier == PackageIdentifier::Eigen
+  && (::pressio::is_native_container_eigen<A_type>::value
+   || ::pressio::is_expression_acting_on_eigen<A_type>::value)
+  && (::pressio::is_vector_eigen<x_type>::value
+   || ::pressio::is_expression_acting_on_eigen<x_type>::value)
+  && (::pressio::is_vector_eigen<y_type>::value
+   || ::pressio::is_expression_acting_on_eigen<y_type>::value)
```
@fnrizzi do you think we could also have `is_container_eigen = is_native_container_eigen || is_expression_acting_on_eigen` convenience shortcut (and similar for Kokkos) to get rid of `and`/`or` mixtures in places like this ?

fnrizzi replied:
i would prefer to wait on that because we need to think abut what it means to be "container eigen".

fnrizzi commented on include/pressio/utils/logger/spdlog/sinks/systemd_sink.hpp:
```diff
@@ -79,7 +79,7 @@ class systemd_sink : public base_sink<Mutex>
     int syslog_level(level::level_enum l)
     {
-        return syslog_levels_.at(static_cast<levels_array::size_type>(l));
+        return syslog_levels_.at(static_cast<levels_array::ordinal_type>(l));
```
why this? this should have nothing to do with traits

mzuzek replied:
Thanks for catching this one! Fixed on e08a2ff261e6b814dac417be7799555ef336d618.

Should this file be in [`unused`](https://github.com/Pressio/pressio/tree/356-new-container-traits/include/pressio/utils/logger/spdlog/sinks/unused) folder ? It doesn't seem to be included anywhere.

fnrizzi replied:
at some point we need to revise all the logging but that should be done independently

fnrizzi commented on include/pressio/utils/logger/spdlog/sinks/syslog_sink.hpp:
```diff
@@ -85,7 +85,7 @@ class syslog_sink : public base_sink<Mutex>
     int syslog_prio_from_level(const details::log_msg &msg) const
     {
-        return syslog_levels_.at(static_cast<levels_array::size_type>(msg.level));
+        return syslog_levels_.at(static_cast<levels_array::ordinal_type>(msg.level));
```
same as other other comment

mzuzek replied:
see [above](https://github.com/Pressio/pressio/pull/418#discussion_r992830610)

fnrizzi commented on include/pressio/ops/ops_ordinal_type.hpp:
```diff
@@ -0,0 +1,158 @@
+
+// short predicate-like form
+template <class T>
+using ordinal_type = typename ContainerInfo<T>::ordinal_type;
```
I am not in favor of this syntax because this diverges from the typical way to get a type in std which relies on the ```_t``` notation. Usually, one sees things like ```std::remove_cvref_t`` to get the nested type. 
Also, there is no advantage in having a ContainerInfo. 

So I would like please to change this so that we can do this: 
```
tempalte<class T, class = void>
struct OrdinalType{
  using type = void;
};

tempalte<class T, class = void>
struct OrdinalType<
  T, ::pressio::mpl::enable_if_t<::pressio::is_expression<T>::value>
  >
{
private:
  using native_type = typename ::pressio::Traits<T>::native_expr_type;
public:
  using type = typename ContainerInfo<native_type>::ordinal_type;
};

template <class T>
using ordinal_type_t = typename OrdinalType<T>::type;
```

And similarly for local ordinal and global ordinal.

fnrizzi commented on include/pressio/ops/tpetra/ops_level2.hpp:
```diff
@@ -105,8 +105,8 @@ _product_tpetra_mv_sharedmem_vec_kokkos(const alpha_t & alpha,
   // make sure the tpetra mv has same exe space of the kokkos vector wrapper
-  using tpetra_mv_dev_t = typename ::pressio::Traits<A_type>::device_type;
-  using kokkos_v_dev_t  = typename ::pressio::Traits<x_type>::device_type;
+  using tpetra_mv_dev_t = typename ::pressio::impl::device_type<A_type>::type;
```
same comment as ordinal_type

mzuzek replied:
Fixed on 9a93a2094d546a7cca5aae6498ae9714f238538b.

fnrizzi commented:
@mzuzek i think after fixing my last comments we should be good to go. but one other thing: can you please verify the tests and ensure we are testing the traits properly?

mzuzek commented:
@fnrizzi I've reviewed the tests and added negative identification checks on https://github.com/Pressio/pressio/pull/418/commits/7b8a378fd8656ed79602ff3b080396566c10361e (with `is_dense_col_major_matrix_eigen` from fd27a67e94a5bc76a54a154a3127d16837924971). Also rebased the branch on current `develop`.

fnrizzi commented on tests/functional_small/type_traits/traits_eigen.cc:
```diff
@@ -7,15 +7,18 @@ namespace pressio { namespace traits { namespace test {
 template <typename T, int rank, bool is_dynamic>
```
if is_dynamic not used , can you remove it?

mzuzek replied:
thanks for catching this - fixed on a4f80c147136f15c9f01d141523877a5dd68e153

fnrizzi replied:
thanks!

fnrizzi commented on tests/functional_small/type_traits/traits_shared.hpp:
```diff
@@ -45,68 +45,88 @@
+}
+
+template <typename T>
+void test_is_not_tpetra_container() {
```
can you separate tpetra from tpetra_block?

mzuzek replied:
sure - fixed on dd029258dbe7be40c965dced2f9577a36e1dcaf4
